### Basic working of Google Palm LLM in LangChain

In [1]:
from langchain.llms import GooglePalm

api_key = 'AIzaSyC-XpTw5ya5fAieMLxGvXz_7UBWcuOquGk' # get this free api key from https://makersuite.google.com/

llm = GooglePalm(google_api_key=api_key, temperature=0.6)

In [2]:
poem = llm("Write a 4 line poem of my love ")
print(poem)

**My love for you is like a red, red rose**
**That's newly sprung in June:**
**My love is like the melody**
**That's sweetly played in tune.**


In [3]:
essay = llm("write email requesting refund for electronic item")
print(essay)

Dear [Name of Customer Service Representative],

I am writing to request a refund for the [product name] that I purchased on [date]. I am not satisfied with the product because [explain why you are not satisfied].

I have attached a copy of my receipt and a picture of the product. I have also included the serial number of the product.

I would like to request a full refund for the product. I would appreciate it if you could process the refund as soon as possible.

Thank you for your time and consideration.

Sincerely,
[Your Name]


In [4]:
from langchain.chains import RetrievalQA


from langchain.embeddings import GooglePalmEmbeddings
from langchain.llms import GooglePalm

### Now let's load data from Sourcegraph FAQ csv file

In [44]:
from langchain.document_loaders.csv_loader import CSVLoader

loader = CSVLoader(file_path='Q&A_LLM.csv', source_column="prompt")

# Store the loaded data in the 'data' variable
data = loader.load()

### Hugging Face Embeddings

In [45]:
from langchain.embeddings import HuggingFaceInstructEmbeddings

# Initialize instructor embeddings using the Hugging Face model
instructor_embeddings = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-large")

e = instructor_embeddings.embed_query("How can i pay for Cody Pro?")

load INSTRUCTOR_Transformer
max_seq_length  512


In [12]:
len(e)

768

In [13]:
e[:5]

[-0.04829464480280876,
 -0.012129939161241055,
 -0.011484012939035892,
 -0.03236175328493118,
 0.03637665882706642]

As you can see above, embedding for a sentence "How can i pay for Cody Pro?" is a list of size 768. Looking at the numbers in this list, doesn't give any intuitive understanding of what it is but just assume that these numbers are capturing the meaning of "How can i pay for Cody Pro?".

### Vector store using FAISS

In [46]:
from langchain.vectorstores import FAISS

# Create a FAISS instance for vector database from 'data'
vectordb = FAISS.from_documents(documents=data,
                                 embedding=instructor_embeddings)

# Create a retriever for querying the vector database
retriever = vectordb.as_retriever(score_threshold = 0.7)

In [40]:
rdocs = retriever.get_relevant_documents("can we buy cody and code search seperate?")
rdocs

[Document(page_content='prompt: can we buy cody and code search seperate?\nresponse: Certainly! There is indeed a different package available. $9 for Cody and $49 code search', metadata={'source': 'can we buy cody and code search seperate?', 'row': 13}),
 Document(page_content='prompt: can you provide both cody and code search?\nresponse: Certainly! There is indeed a bundled package available. $ 59', metadata={'source': 'can you provide both cody and code search?', 'row': 12}),
 Document(page_content='prompt: can i run both Code search and Cody enterprise?\nresponse: You can purchase both Code Search Enterprise and Cody Enterprise and run them off of the same Sourcegraph deployment. Contact us to learn more.', metadata={'source': 'can i run both Code search and Cody enterprise?', 'row': 19}),
 Document(page_content="prompt: What's the difference between the Cody Free, Pro, nad Enterprise plans?\nresponse: Cody Free and Cody Pro are both offered as standalone products. Cody Free is best

As you can see above, the retriever that was created using FAISS and hugging face embedding is now capable of pulling relavant documents from our original CSV file knowledge store. This is very powerful and it will help us further in our project

##### Embeddings can be created using GooglePalm too. Also for vector database you can use chromadb as well as shown below. During our experimentation, we found hugging face embeddings and FAISS to be more appropriate for our use case

In [3]:
# google_palm_embeddings = GooglePalmEmbeddings(google_api_key=api_key)

# from langchain.vectorstores import Chroma
# vectordb = Chroma.from_documents(data,
#                            embedding=google_palm_embeddings,
#                            persist_directory='./chromadb')
# vectordb.persist()

### Create RetrievalQA chain along with prompt template 🚀

In [47]:
from langchain.prompts import PromptTemplate

prompt_template = """Given the following context and a question, generate an answer based on this context only.
In the answer try to provide as much text as possible from "response" section in the source document context without making much changes.
If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

CONTEXT: {context}

QUESTION: {question}"""


PROMPT = PromptTemplate(
    template=prompt_template, input_variables=["context", "question"]
)
chain_type_kwargs = {"prompt": PROMPT}


from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",
                            retriever=retriever,
                            input_key="query",
                            return_source_documents=True,
                            chain_type_kwargs=chain_type_kwargs)


### We are all set 👍🏼 Let's ask some questions now

In [26]:
chain('I have zero knowledge of AI and belong to a non-technical, can i able to use?')

{'query': 'I have zero knowledge of AI and belong to a non-technical, can i able to use?',
 'result': 'Yes, you can certainly use AI technologies even if you have zero technical knowledge',
 'source_documents': [Document(page_content='prompt: I have zero knowledge of AI and belong to a non-technical, can i able to use?\nresponse: Yes, you can certainly use AI technologies even if you have zero technical knowledge', metadata={'source': 'I have zero knowledge of AI and belong to a non-technical, can i able to use?', 'row': 24}),
  Document(page_content='prompt: I have never used AI assistant in my life. Can I take Cody as an AI assistant this?\nresponse: Yes, of course! You can consider Cody as your AI assistant for this task.', metadata={'source': 'I have never used AI assistant in my life. Can I take Cody as an AI assistant this?', 'row': 14}),
  Document(page_content='prompt: Is there any prerequisite for buying this AI assistant ?\nresponse: No prerequisites are necessary for purchas

**As you can see above, the answer of question comes from four different FAQs within our csv file and it is able to pull those questions and merge them nicely**

In [27]:
chain("How can i pay for the code Intelligence Platform plan?")

{'query': 'How can i pay for the code Intelligence Platform plan?',
 'result': 'The Code Intelligence Platform plan requires an annual contract. Contact us to learn more.',
 'source_documents': [Document(page_content='prompt: How can i pay for the code Intelligence Platform plan?\nresponse: The Code Intelligence Platform plan requires an annual contract. Contact us to learn more.', metadata={'source': 'How can i pay for the code Intelligence Platform plan?', 'row': 22}),
  Document(page_content='prompt: how can i pay for code search?\nresponse: Code Search requires an annual contract. Contact us to learn more.', metadata={'source': 'how can i pay for code search?', 'row': 20}),
  Document(page_content='prompt: How can i pay for Cody Pro?\nresponse: Cody Pro is billed on a monthly basis and can be paid with a credit card.', metadata={'source': 'How can i pay for Cody Pro?', 'row': 4}),
  Document(page_content='prompt: what is the code intelligence Platform?\nresponse: The Code Intellige

In [30]:
chain("Does Sourcegraph work in Mac OS/Ubuntu?")

{'query': 'Does Sourcegraph work in Mac OS/Ubuntu?',
 'result': 'Yes',
 'source_documents': [Document(page_content='prompt: Does Sourcegraph work in Mac OS/Ubuntu?\nresponse: Yes', metadata={'source': 'Does Sourcegraph work in Mac OS/Ubuntu?', 'row': 29}),
  Document(page_content='prompt: how and where can i install sourcegraph?\nresponse: Visit the Sourcegraph website and navigate to the installation section, Download the appropriate installer for your operating system (e.g., Windows, macOS, Linux), and Follow the installation instructions provided on the website or in the downloaded package. Once installed, configure Sourcegraph settings as needed, Access Sourcegraph through your web browser by entering the specified URL or launching it from your applications menu.', metadata={'source': 'how and where can i install sourcegraph?', 'row': 31}),
  Document(page_content='prompt: is my data secure when connected to code search?\nresponse: Sourcegraph has security and reliability controls 

In [29]:
chain("how and where can i install sourcegraph?")

{'query': 'how and where can i install sourcegraph?',
 'result': 'Visit the Sourcegraph website and navigate to the installation section, Download the appropriate installer for your operating system (e.g., Windows, macOS, Linux), and Follow the installation instructions provided on the website or in the downloaded package. Once installed, configure Sourcegraph settings as needed, Access Sourcegraph through your web browser by entering the specified URL or launching it from your applications menu.',
 'source_documents': [Document(page_content='prompt: how and where can i install sourcegraph?\nresponse: Visit the Sourcegraph website and navigate to the installation section, Download the appropriate installer for your operating system (e.g., Windows, macOS, Linux), and Follow the installation instructions provided on the website or in the downloaded package. Once installed, configure Sourcegraph settings as needed, Access Sourcegraph through your web browser by entering the specified URL 

In [31]:
chain("Can you explain more about your product?")

{'query': 'Can you explain more about your product?',
 'result': 'Yes, Please Visit our website (https://sourcegraph.com)',
 'source_documents': [Document(page_content='prompt: Can you explain more about your product?\nresponse: Yes, Please Visit our website (https://sourcegraph.com)', metadata={'source': 'Can you explain more about your product?', 'row': 37}),
  Document(page_content='prompt: How many products you own?\nresponse: there are 3 products', metadata={'source': 'How many products you own?', 'row': 10}),
  Document(page_content='prompt: How can I get help if I have a doubt and need support?\nresponse: Contact us', metadata={'source': 'How can I get help if I have a doubt and need support?', 'row': 26}),
  Document(page_content="prompt: Will there be an upgraded when there are new features in Sourcegraph?\nresponse: Yes, we'll let you know!", metadata={'source': 'Will there be an upgraded when there are new features in Sourcegraph?', 'row': 27})]}

In [32]:
chain("Once purchased, is this course available for lifetime access?")

{'query': 'Once purchased, is this course available for lifetime access?',
 'result': 'Contact us',
 'source_documents': [Document(page_content='prompt: Once purchased, is this course available for lifetime access?\nresponse: Contact us', metadata={'source': 'Once purchased, is this course available for lifetime access?', 'row': 25}),
  Document(page_content='prompt: is this free?\nresponse: Yes, it comes with a single developer limit and restricted access.', metadata={'source': 'is this free?', 'row': 11}),
  Document(page_content='prompt: How can i upgrade or downgrade my plans?\nresponse: Users can upgrade or downgrade between Cody Free and Cody Pro within their account settings at any time. To upgrade to Cody Enterprise, please contact our Sales team.', metadata={'source': 'How can i upgrade or downgrade my plans?', 'row': 5}),
  Document(page_content='prompt: How can i pay for the code Intelligence Platform plan?\nresponse: The Code Intelligence Platform plan requires an annual co

In [35]:
chain("How can i upgrade or downgrade my plans?")

{'query': 'How can i upgrade or downgrade my plans?',
 'result': 'Users can upgrade or downgrade between Cody Free and Cody Pro within their account settings at any time. To upgrade to Cody Enterprise, please contact our Sales team.',
 'source_documents': [Document(page_content='prompt: How can i upgrade or downgrade my plans?\nresponse: Users can upgrade or downgrade between Cody Free and Cody Pro within their account settings at any time. To upgrade to Cody Enterprise, please contact our Sales team.', metadata={'source': 'How can i upgrade or downgrade my plans?', 'row': 5}),
  Document(page_content='prompt: How can i pay for the code Intelligence Platform plan?\nresponse: The Code Intelligence Platform plan requires an annual contract. Contact us to learn more.', metadata={'source': 'How can i pay for the code Intelligence Platform plan?', 'row': 22}),
  Document(page_content='prompt: What forms of support are available for paid Cody plans?\nresponse: Email and web portal support is

In [43]:
chain("is my data secure when connected to code search?")

{'query': 'is my data secure when connected to code search?',
 'result': 'Sourcegraph has security and reliability controls built for the most demanding enterprises. To learn more, see our Security page.',
 'source_documents': [Document(page_content='prompt: is my data secure when connected to code search?\nresponse: Sourcegraph has security and reliability controls built for the most demanding enterprises. To learn more, see our Security page.', metadata={'source': 'is my data secure when connected to code search?', 'row': 18}),
  Document(page_content='prompt: How do i deploy Sourcegraph Code Search? Can I run it self-hosted\nresponse: Code Search can be deployed using Sourcegraph Cloud (we host) or Sourcegraph Self-hosted (you host).', metadata={'source': 'How do i deploy Sourcegraph Code Search? Can I run it self-hosted', 'row': 17}),
  Document(page_content='prompt: how can i pay for code search?\nresponse: Code Search requires an annual contract. Contact us to learn more.', metad

In [49]:
chain("What's the difference between the Cody Free, Pro, and Enterprise plans?")

{'query': "What's the difference between the Cody Free, Pro, and Enterprise plans?",
 'result': "Cody Free and Cody Pro are both offered as standalone products. Cody Free is best for individuals working on hobby projects or professional devs trying out AI coding assistants. Usage limits apply.  Cody Pro is best for professional developers who want unlimited usage and context for larger codebases, and Cody Enterprise is for organizations needing security, scalability, and control. It is coming soon for everyone, but is currently available for existing Code Search customers (Contact us if you'd like to learn more).",
 'source_documents': [Document(page_content="prompt: What's the difference between the Cody Free, Pro, and Enterprise plans?\nresponse: Cody Free and Cody Pro are both offered as standalone products. Cody Free is best for individuals working on hobby projects or professional devs trying out AI coding assistants. Usage limits apply.  Cody Pro is best for professional develope